File: backend/app/prng.py
"""A pseudo-random generator written out in full, so that two languages can agree.

The colony's ground is generated on the server, by this code, and in the browser, by its twin
in TypeScript. That is only sound if both produce the same numbers -- and `random.Random` is
CPython's Mersenne Twister with CPython's `gauss`, which would have to be reimplemented in
JavaScript and kept correct for ever. Reimplementing a language's standard library is the
wrong end to start from.

So the algorithm is made portable instead. Mulberry32 is thirty-two-bit integer arithmetic and
nothing else: `Math.imul` in JavaScript and a mask in Python compute it identically, with no
appeal to a floating-point library and no per-implementation choices to get wrong.

Directions come out of sphere point picking rather than out of normalised gaussians. Partly
because it is exactly uniform where normalising three gaussians only approximately is, and
partly because it needs no logarithm: every transcendental call is one more place where two
libm implementations can disagree in the last bit.
"""
from __future__ import annotations

import math

MASK = 0xFFFFFFFF


def seed_int(text: str) -> int:
    """FNV-1a, 32 bit. Spelled out for the same reason as the rest of this module."""
    h = 2166136261
    for byte in text.encode("utf-8"):
        h = ((h ^ byte) * 16777619) & MASK
    return h


class Prng:
    """Mulberry32. One 32-bit word of state, and every step is a 32-bit integer operation."""

    __slots__ = ("state",)

    def __init__(self, seed: int | str):
        self.state = seed_int(seed) if isinstance(seed, str) else seed & MASK

    def next_uint(self) -> int:
        self.state = (self.state + 0x6D2B79F5) & MASK
        t = self.state
        t = ((t ^ (t >> 15)) * (t | 1)) & MASK
        t = (t ^ (t + ((t ^ (t >> 7)) * (t | 61)) & MASK)) & MASK
        return (t ^ (t >> 14)) & MASK

    def random(self) -> float:
        return self.next_uint() / 4294967296.0

    def uniform(self, low: float, high: float) -> float:
        return low + (high - low) * self.random()

    def randint(self, low: int, high: int) -> int:
        """Inclusive at both ends, like `random.randint`."""
        return low + int(self.random() * (high - low + 1))

    def direction(self) -> tuple[float, float, float]:
        """A point picked uniformly on the unit sphere."""
        z = 2.0 * self.random() - 1.0
        phi = 2.0 * math.pi * self.random()
        r = math.sqrt(max(0.0, 1.0 - z * z))
        return (r * math.cos(phi), r * math.sin(phi), z)

File: frontend/app/colony/prng.ts
// The twin of `backend/app/prng.py`. Both produce the same numbers, and
// `citygen.test.ts` holds them to it against a reference the Python side wrote.
//
// Mulberry32: one word of state, thirty-two-bit integer arithmetic and nothing else. Chosen
// over reimplementing CPython's Mersenne Twister and its `gauss`, which is where this would
// otherwise have had to start -- reimplementing a language's standard library is the wrong
// end to pick up.

const MASK = 0xffffffff;

/** FNV-1a, 32 bit. Spelled out, like the rest of this file. */
export function seedInt(text: string): number {
  let h = 2166136261 >>> 0;
  const bytes = new TextEncoder().encode(text);
  for (const byte of bytes) h = Math.imul(h ^ byte, 16777619) >>> 0;
  return h >>> 0;
}

export class Prng {
  private state: number;

  constructor(seed: number | string) {
    this.state = (typeof seed === "string" ? seedInt(seed) : seed) >>> 0;
  }

  nextUint(): number {
    this.state = (this.state + 0x6d2b79f5) >>> 0;
    let t = this.state;
    t = Math.imul(t ^ (t >>> 15), t | 1) >>> 0;
    t = (t ^ (t + (Math.imul(t ^ (t >>> 7), t | 61) >>> 0))) >>> 0;
    return (t ^ (t >>> 14)) >>> 0;
  }

  random(): number {
    return this.nextUint() / 4294967296;
  }

  uniform(low: number, high: number): number {
    return low + (high - low) * this.random();
  }

  /** Inclusive at both ends, matching Python's `randint`. */
  randint(low: number, high: number): number {
    return low + Math.floor(this.random() * (high - low + 1));
  }

  /** A point picked uniformly on the unit sphere -- no logarithm, so one fewer place where
   *  two maths libraries can disagree in the last bit. */
  direction(): [number, number, number] {
    const z = 2 * this.random() - 1;
    const phi = 2 * Math.PI * this.random();
    const r = Math.sqrt(Math.max(0, 1 - z * z));
    return [r * Math.cos(phi), r * Math.sin(phi), z];
  }
}
